### PR TITLE
add `ng-select` for templatable select options

### LINKED ISSUE #3015: video privacy text mismatch
**What happened?**
screenshots of "upload" first page. it shows in the privacy, 
 1. public
 2. unlisted
 3. private
 4. internal
 5. scheduled

**Further, "scheduled" option is NOT available in import url or torrent import.** 

Once a video has been uploaded or imported, the "edit a video" page shows the same privacy but now the text has changed, 
it says
1. only i can see this video
2. only people with a private link can see this video
3. anyone can see this video
4. only users of this instance can see this video

**What do you expect to happen instead?**
make the language clear about the privacy. I have not checked the upload on any more updated instance because i havent found one which would let me upload. close the issue if its been fixed in future versions

**Additional information**
* PeerTube version or URL:https://devtube.dev-wiki.de 

PeerTube version | 2.1.1
-- | --

### PR DESCRIPTION
**Pull request by rigelk**
Having a slightly more customizable select component is required here to display both the label of an option and its description. [`ng-select`](https://github.com/ng-select/ng-select) component, like most select components around, features lazy-loading of remote data - which could prove useful for listing channels in the `video-edit` view in the future.

In this PR I focus on the upload/edit view, but other select usages could use option templates elsewhere.

Screenshot_2020-08-05 Upload a video - PeerTube
Screenshot_2020-08-05 Edit a video - PeerTube

fixes #3015

### CODE REVIEW

**Comment by rigelk** [comment]
@Chocobozzz I didn't check bundle size, because I think this component could replace `ngx-tag` as well  and save on its size: https://ng-select.github.io/ng-select/#/tags

Considered this, I didn't try to target the video-update module only.

**Comment by Chocobozzz** [comment]
:heart_eyes:

File: client/package.json [hunk 1]
```diff
@@ -135,5 +135,7 @@
     "whatwg-fetch": "^3.0.0",
     "zone.js": "~0.10.2"
   },
-  "dependencies": {}
+  "dependencies": {
+    "@ng-select/ng-select": "^5.0.0"
```
Comment (Chocobozzz) [inline client/package.json:139]: dev dep

File: client/src/app/+videos/+video-edit/shared/video-caption-add-modal.component.html [hunk 2]
```diff
@@ -8,11 +8,11 @@ <h4 i18n class="modal-title">Add caption</h4>
 
     <div class="modal-body">
       <label i18n for="language">Language</label>
-      <div class="peertube-select-container">
-        <select id="language" formControlName="language" class="form-control">
-          <option></option>
-          <option *ngFor="let language of videoCaptionLanguages" [value]="language.id">{{ language.label }}</option>
-        </select>
+      <div class="peertube-ng-select-container">
```
Comment (Chocobozzz) [inline client/src/app/+videos/+video-edit/shared/video-caption-add-modal.component.html:11]: We still need a container with ng select?

File: client/src/app/+videos/+video-edit/video-add-components/video-import-torrent.component.html [hunk 3]
```diff
@@ -25,19 +25,30 @@
 
     <div class="form-group">
       <label i18n for="first-step-channel">Channel</label>
-      <div class="peertube-select-container">
-        <select id="first-step-channel" [(ngModel)]="firstStepChannelId" class="form-control">
-          <option *ngFor="let channel of userVideoChannels" [value]="channel.id">{{ channel.label }}</option>
-        </select>
+      <div class="peertube-ng-select-container">
+        <ng-select
```
Comment (Chocobozzz) [inline client/src/app/+videos/+video-edit/video-add-components/video-import-torrent.component.html:29]: Could you create a dedicated component to choose user channels?

File: client/src/app/+videos/+video-edit/video-add-components/video-import-url.component.html [hunk 4]
```diff
@@ -20,19 +20,30 @@
...
     </div>
 
     <div class="form-group">
       <label i18n for="first-step-privacy">Privacy</label>
-      <div class="peertube-select-container">
-        <select id="first-step-privacy" [(ngModel)]="firstStepPrivacyId" class="form-control">
-          <option *ngFor="let privacy of videoPrivacies" [value]="privacy.id">{{ privacy.label }}</option>
-        </select>
+      <div class="peertube-ng-select-container">
+        <ng-select
```
Comment (Chocobozzz) [inline client/src/app/+videos/+video-edit/video-add-components/video-import-url.component.html:39]: Could you create a dedicated component for a select + description? (so we just have to send an object)

File: client/src/app/+videos/+video-edit/video-update.resolver.ts [hunk 5]
```diff
@@ -28,7 +28,12 @@ export class VideoUpdateResolver implements Resolve<any> {
                          .listAccountVideoChannels(video.account)
                          .pipe(
                            map(result => result.data),
-                           map(videoChannels => videoChannels.map(c => ({ id: c.id, label: c.displayName, support: c.support })))
+                           map(videoChannels => videoChannels.map(c => ({
+                             id: c.id,
+                             label: c.displayName,
+                             support: c.support,
+                             avatar: c.avatar.path
```
Comment (Chocobozzz) [inline client/src/app/+videos/+video-edit/video-update.resolver.ts:35]: avatar could be null no?

File: client/src/app/helpers/utils.ts [hunk 6]
```diff
@@ -27,7 +30,7 @@ function populateAsyncUserVideoChannels (authService: AuthService, channel: { id
           const videoChannels = user.videoChannels
           if (Array.isArray(videoChannels) === false) return
 
-          videoChannels.forEach(c => channel.push({ id: c.id, label: c.displayName, support: c.support }))
+          videoChannels.forEach(c => channel.push({ id: c.id, label: c.displayName, support: c.support, avatar: c.avatar?.path }))
```
Comment (Chocobozzz) [inline client/src/app/helpers/utils.ts:33]: `avatarPath: c.avatar?.path`

File: client/src/app/shared/shared-main/video/video.service.ts [hunk 7]
```diff
@@ -339,23 +339,25 @@ export class VideoService implements VideosProvider {
...
         id: VideoPrivacy.INTERNAL,
-        label: this.i18n('Only users of this instance can see this video')
+        description: this.i18n('Only users of this instance can see this video')
       }
     ]
 
-    return base.filter(o => !!privacies.find(p => p.id === o.id))
+    return base
+      .filter(o => !!privacies.find(p => p.id === o.id))
+      .map(o => ({ ...privacies[o.id - 1], ...o }))
```
Comment (Chocobozzz) [inline client/src/app/shared/shared-main/video/video.service.ts:360]: wut o_O

File: client/package.json (hunk 1 above)
Comment (rigelk) [reply client/package.json:139]: ofc, I forgot even the basics ><

File: client/src/app/+videos/+video-edit/shared/video-caption-add-modal.component.html (hunk 2 above)
Comment (rigelk) [reply client/src/app/+videos/+video-edit/shared/video-caption-add-modal.component.html:11]: better safe than sorry ^^

File: client/src/app/shared/shared-main/video/video.service.ts (hunk 7 above)
Comment (rigelk) [reply client/src/app/shared/shared-main/video/video.service.ts:360]: lemme comment that

File: client/src/app/+videos/+video-edit/video-add-components/video-import-torrent.component.html (hunk 3 above)
Comment (rigelk) [reply client/src/app/+videos/+video-edit/video-add-components/video-import-torrent.component.html:29]: @Chocobozzz should I put that in `client/src/app/shared/shared-forms/select-{channel,options}`?

Comment (Chocobozzz) [reply client/src/app/+videos/+video-edit/video-add-components/video-import-torrent.component.html:29]: Yep

**Comment by rigelk** [comment]
@Chocobozzz I made progress, but I also have a strange error that might be due to my modifications of the `video-edit` view:

Peek 2020-08-07 10-20 - problem with ng-select

File: client/src/app/+videos/+video-edit/shared/video-caption-add-modal.component.html (hunk 2 above)
Comment (rigelk) [reply client/src/app/+videos/+video-edit/shared/video-caption-add-modal.component.html:11]: I removed it, now that components are wrapping it anyway.

**Comment by rigelk** [comment]
@Chocobozzz I replaced `ngx-chips` as intended, and while my `select-tags` component behaves well in forms (in the `video-edit` view), I can't seem to make it work in the search-filter component, where there is no form and two-way data binding doesn't seem to work with it. I could always replace the  component with the bare `ng-select` component, but it kinda defeats the factorization.

File: client/src/app/shared/shared-forms/select-tags.component.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import { Component, Input, forwardRef } from '@angular/core'
+import { NG_VALUE_ACCESSOR, ControlValueAccessor } from '@angular/forms'
+
+@Component({
+  selector: 'select-tags',
```
Comment (Chocobozzz) [inline client/src/app/shared/shared-forms/select-tags.component.ts:5]: `my-select-tags`

File: client/src/app/shared/shared-forms/select-options.component.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
+import { Component, Input, forwardRef } from '@angular/core'
+import { NG_VALUE_ACCESSOR, ControlValueAccessor } from '@angular/forms'
+
+type Item = { id: number | string, label: string, description?: string }
+
+@Component({
+  selector: 'select-options',
```
Comment (Chocobozzz) [inline client/src/app/shared/shared-forms/select-options.component.ts:7]: `my-select-options`

File: client/src/app/shared/shared-forms/select-options.component.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
+import { Component, Input, forwardRef } from '@angular/core'
+import { NG_VALUE_ACCESSOR, ControlValueAccessor } from '@angular/forms'
+
+type Item = { id: number | string, label: string, description?: string }
```
Comment (Chocobozzz) [inline client/src/app/shared/shared-forms/select-options.component.ts:4]: `SelectOptionsItem` and export it so parent components can directly use this type

File: client/src/app/shared/shared-forms/select-channel.component.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
+import { NgSelectComponent } from '@ng-select/ng-select'
+import { Component, Input, forwardRef, ViewChild } from '@angular/core'
+import { NG_VALUE_ACCESSOR, ControlValueAccessor } from '@angular/forms'
+import { Actor } from '../shared-main'
+
+@Component({
+  selector: 'select-channel',
```
Comment (Chocobozzz) [inline client/src/app/shared/shared-forms/select-channel.component.ts:7]: `my-select-channel`

File: client/src/app/shared/shared-forms/select-channel.component.html [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+  [bindLabel]="bindLabel"
+  [bindValue]="bindValue"
+  [clearable]="clearable"
+  [searchable]="searchable"
+>
+  <ng-option *ngFor="let channel of channels" [value]="channel.id">
+    <img
+      class="avatar mr-1"
+      [src]="channel.avatarPath"
+      height="20px" width="20px"
```
Comment (Chocobozzz) [inline client/src/app/shared/shared-forms/select-channel.component.html:13]: better to use CSS for `height` and `width` no?

File: client/src/app/shared/shared-forms/select-channel.component.html [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+  [searchable]="searchable"
+>
+  <ng-option *ngFor="let channel of channels" [value]="channel.id">
+    <img
+      class="avatar mr-1"
+      [src]="channel.avatarPath"
+    />
+    {{ channel.label }}
+  </ng-option>
+</ng-select>
```
Comment (kimsible) [inline client/src/app/shared/shared-forms/select-channel.component.html:16]: missed new line ?

File: client/src/app/shared/shared-forms/select-options.component.html [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+  [searchable]="searchable"
+>
+  <ng-template ng-option-tmp let-item="item" let-index="index">
+    {{ item.label }}
+    <ng-container *ngIf="item.description">
+      <br>
+      <span [title]="item.description" class="text-muted">{{ item.description }}</span>
+    </ng-container>
+  </ng-template>
+</ng-select>
```
Comment (kimsible) [inline client/src/app/shared/shared-forms/select-options.component.html:18]: a missing new line here too

**Pull request merged**